specification/dsdl/architecture.tex

```diff
@@ -0,0 +1,233 @@
+
+        \texttt{\qquad{}\qquad{}\qquad{}102.List.2.0.uavcan} &
+        Data type definition v2.0 with fixed service ID 102. \\\cline{2-2}
+
+        \texttt{\qquad{}\qquad{}\qquad{}ID.1.0.uavcan} &
+        Data type definition v1.0 without fixed port ID. \\\hline
+    \end{tabu}
+    \caption{DSDL directory structure example.}\label{fig:dsdl_directory_structure_example}
+\end{figure}
+
+\subsection{Elements of data type definition}
+
+A data type definition file contains an exhaustive description of a particular version of the said data type in the
+\emph{data structure description language} (DSDL).
+As explained above, one data type definition contains either one or two data schema definitions,
+for message types and service types, respectively.
+
+The term \emph{attribute} is used to refer to either a field (including padding fields) or a constant
+associated with a particular object or type.
+
+A data schema definition contains an ordered, possibly empty collection of \emph{fields} and/or
+unordered, possibly empty collection of \emph{constants}.
+A data schema may describe either a \emph{structure object} or a \emph{tagged union object}.
+The value of a structure object is a function of the values of all of its fields.
+A tagged union object is formed from at least two fields,
+but it is capable of holding exactly one field at any given time.
+The value of a tagged union object is a function of which field it is holding at the moment and the value of said field.
+
+A field represents a named dynamically assigned value of a statically defined type
+that can be exchanged over the network as a member of its containing object.
+A padding field is a special kind of field which is used for data alignment purposes;
+such fields are not named.
+
+A constant represents a named statically defined value of a statically defined type.
+Constants are never exchanged over the network, since they are assumed to be known to all involved nodes
+by virtue of them sharing the same static definition of the data type.
+
+Constant values are defined via \emph{DSDL expressions},
+which are evaluated at the time of DSDL definition processing.
+There is a special category of types called \emph{expression types},
+instances of which are used only during expression evaluation
+and cannot be exchanged over the network.
+
+Data type definitions can also contain various auxiliary elements reviewed later,
+such as deprecation markers (notifying its users that the data type is no longer recommended for new designs)
+or assertions (special statements introduced by data type designers
+which are statically validated by DSDL processing tools).
+
+\subsection{Serialization}
+
+Every object that can be exchanged between UAVCAN nodes has a well-defined \emph{serialized representation}.
+The value and meaning of an object can be unambiguously recovered from its serialized representation,
+provided that the type of the object is known.
```

Comment:
We need a section describing how types are "known". This is especially significant now that we no longer have "message ids" but are using port ID instead. With this we've weakened the link between a CAN identifier and a DSDL type making it possible, for example, to exchange the same type over different ports. We've talked about how such mappings will have to be installed into a system but I'm not sure we've discussed this in the document? If we have, sorry, just add cross-reference to this discussion.

Reply:
This is very valid. Please review this commit I just pushed here: https://github.com/UAVCAN/specification/pull/46/commits/33b69a58ff9e0aff8117d404a143f9ea26042e09